partiql-planner/src/test/kotlin/org/partiql/planner/TyperTestBuilder.kt

```diff
@@ -0,0 +1,196 @@
+ * this will either return a [TyperTest.Success] or [TyperTest.Failure].
+ * @see FIELD_STATUS
+ * @see STATUS_SUCCESS
+ * @see STATUS_FAILURE
+ * @see TyperTest
+ * @see TyperTest.Success
+ * @see TyperTest.Failure
+ */
+class TyperTestBuilder : TestBuilder {
+
+    private var name: TestId? = null
+    private var expectedType: PType? = null
+    private var isSuccess: Boolean = true
+    private var statement: String? = null
+    private var sessionCatalog: String? = null
+    private var sessionDirectory: List<String>? = null
+    private var inputProvider: PartiQLTestProvider? = null
+    private var assertion: ProblemAssertion? = null
+
+    companion object {
+        const val STATUS_SUCCESS = "SUCCESS"
+        const val STATUS_FAILURE = "FAILURE"
+        const val FIELD_EXPECTED = "expected"
+        const val FIELD_STATUS = "status"
+        const val FIELD_STATEMENT = "statement"
+        const val FIELD_SESSION = "session"
+        const val FIELD_CATALOG = "catalog"
+        const val FIELD_CWD = "cwd"
+        const val FIELD_ASSERT_PROBLEM_EXISTS = "assertProblemExists"
+    }
+
+    override fun id(id: TestId): TestBuilder = this.apply {
+        this.name = id
+    }
+
+    override fun config(body: StructElement, statementProvider: PartiQLTestProvider): TestBuilder = this.apply {
+        this.inputProvider = statementProvider
+        val session = body[FIELD_SESSION].asStruct()
+        this.sessionCatalog = session[FIELD_CATALOG].asString().textValue
+        this.sessionDirectory = session[FIELD_CWD].asList().values.map { it.asString().textValue }
+        this.statement = getStatement(body[FIELD_STATEMENT])
+        val status = body[FIELD_STATUS].asSymbol().textValue
+        when (status) {
+            STATUS_SUCCESS -> {
+                this.isSuccess = true
+                this.expectedType = TypeParser.parse(body[FIELD_EXPECTED])
+            }
+            STATUS_FAILURE -> {
+                this.isSuccess = false
+                if (body.containsField(FIELD_EXPECTED)) {
+                    this.expectedType = TypeParser.parse(body[FIELD_EXPECTED])
+                }
+                if (body.containsField(FIELD_ASSERT_PROBLEM_EXISTS)) {
+                    val sexp = body[FIELD_ASSERT_PROBLEM_EXISTS].asSexp()
+                    val severity = sexp.values[0].textValue
+                    val message = sexp.values[1].textValue
+                    this.assertion = ProblemAssertion(severity, message)
+                }
+            }
+        }
+    }
+
+    override fun build(): Test {
+        return when (this.isSuccess) {
+            true -> TyperTest.Success(this.name.toString(), this.expectedType!!, this.statement!!, this.sessionCatalog!!, this.sessionDirectory!!)
+            false -> TyperTest.Failure(this.name.toString(), this.expectedType, this.assertion, this.statement!!, this.sessionCatalog!!, this.sessionDirectory!!)
```

Comment:
Could we make more of these vars be non-nullable? Some of these will never be `null` post-initialization such as `statement` and `sessionCatalog`.